src/providers/pkcs11/mod.rs

```diff
@@ -479,27 +487,75 @@ impl ProviderBuilder {
                 Error::new(ErrorKind::InvalidData, "error initializing PKCS 11 context")
             })?;
 
-        let slot_number = match self.slot_number {
-            Some(i) => {
-                let slot = Slot::try_from(i).or_else(|_| {
+        let slots = backend.get_slots_with_initialized_token().map_err(|e| {
+            format_error!(
+                "Failed retrieving a valid slot with an initialized token",
+                e
+            );
+            Error::new(
+                ErrorKind::InvalidData,
+                "Failed retrieving a valid slot with an initialized token",
+            )
+        })?;
+        let slot_number = match (self.serial_number, self.slot_number) {
+            (Some(serial_number), _given_slot) => {
```

Comment:
The `_given_slot` name shouldn't start with `_` if you actually use it below, just `given_slot` should do.